docs/guide/custom-ui.md

```diff
@@ -0,0 +1,207 @@
+
+The `AposWidgetEditor` component already provides a modal dialog box in which to edit the schema of any widget, so we won't need to configure a replacement unless we want to support editing directly on the page. `AposRichTextWidgetEditor` is an example of how to do this.
+
+The `AposWidget` component has **nothing to do with a typical site visitor experience.** It is used only when displaying our widget while the page is in edit mode. While overriding this component is rare, but the `@apostrophecms/rich-text-widget` module does so to provide a "click the text to edit" experience for rich text widgets. If you're just trying to enhance your widget with frontend JavaScript, you should write a [widget player](custom-widgets.md#client-side-javascript-for-widgets) instead.
+:::
+
+## Registering custom field types
+
+Apostrophe's [schema field types](content-schema.md) cover many situations, but we might wish to add a new one.
+
+A schema field has two parts: a server-side part and a browser-side part. The server-side part is responsible for sanitizing the input received from the browser, while the browser-side part is responsible for providing the admin UI.
+
+### Implementing the server-side part
+
+Any module can register a schema field type on the server side, like this one, which provides a range field with two inputs, one for the low end of the range and one for the high end.
+
+<AposCodeBlock>
+```js
+module.exports = {
+  init(self) {
+    self.addStarRatingFieldType();
+  },
+  methods(self) {
+    return {
+      addStarRatingFieldType() {
+        self.apos.schema.addFieldType({
+          name: 'starRating',
+          convert: self.convert,
+          vueComponent: 'InputStarRating'
+        });
+      },
+      async convert(req, field, data, object) {
+        const input = data[field.name];
+        if ((data[field.name] == null) || (data[field.name] === '')) {
+          if (field.required) {
+            throw self.apos.error('notfound');
+          }
+        }
+        object[field.name] = self.apos.launder.integer(input, field.def, 1, 5);
+      }
+    }
+  }
+}
+```
+  <template v-slot:caption>
+    modules/star-rating-field/index.js
+  </template>
+</AposCodeBlock>
+
+In `init`, which runs when the module is initialized, we call our `addStarRatingFieldType` method. `init` is the right place to invoke code that should run when the Apostrophe process starts up.
+
+In `addStarRatingFieldType`, we invoke `self.apos.schema.addFieldType` to add our custom field type on the server side. We provide:
+
+* `name`, which can be used as a `type` setting when adding the field to a schema.
+* `convert`, a function to be used to sanitize the input and copy it to a destination. We pass our `convert` method for this purpose. Methods of our module are available as properties of `self`.
```

Comment:
> We pass our `convert` method for this purpose.

This took me a minute to work out since it's in the description of a `convert` function. Maybe we should name the method something slightly different for clarity.